src/empirical.cpp

```diff
@@ -0,0 +1,85 @@
+  vector<double> theta_v = as_vector(theta);
+  matrix<double> fitted = x_matrix * beta_m;
+  matrix<double> y_matrix = as_vector(y).matrix();
+  matrix<double> residual = y_matrix - fitted;
+  vector<double> G_sqrt = as_vector(sqrt(weights_vector));
+  int n_theta = theta.size();
+  int theta_size_per_group = n_theta / n_groups;
+  int p = x.cols();
+  // Use map to hold these base class pointers (can also work for child class objects).
+  std::map<int, derivatives_base<double>*> derivatives_by_group;
+  for (int r = 0; r < n_groups; r++) {
+    // in loops using new keyword is required so that the objects stays on the heap
+    // otherwise this will be destroyed and you will get unexpected result
+    if (is_spatial) {
+      derivatives_by_group[r] = new derivatives_sp_exp<double>(vector<double>(theta_v.segment(r * theta_size_per_group, theta_size_per_group)));
+    } else {
+      derivatives_by_group[r] = new derivatives_nonspatial<double>(vector<double>(theta_v.segment(r * theta_size_per_group, theta_size_per_group)), n_visits, cov_type);
+    }
+  }
+  matrix<double> meat = matrix<double>::Zero(beta_vcov_matrix.rows(), beta_vcov_matrix.cols());
+  
+  for (int i = 0; i < n_subjects; i++) {
+    int start_i = subject_zero_inds[i];
+    int n_visits_i = subject_n_visits[i];
+    std::vector<int> visit_i(n_visits_i);
+    matrix<double> dist_i(n_visits_i, n_visits_i);
+    if (!is_spatial) {
+      for (int i = 0; i < n_visits_i; i++) {
+        visit_i[i] = visits_zero_inds[i + start_i];
+      }
+    } else {
+      dist_i = euclidean(matrix<double>(coords.block(start_i, 0, n_visits_i, coordinates.cols())));
+    }
+    int subject_group_i = subject_groups[i] - 1;
+    matrix<double> sigma_inv_chol = derivatives_by_group[subject_group_i]->get_inverse_chol(visit_i, dist_i);
+    matrix<double> Xi = x_matrix.block(start_i, 0, n_visits_i, x_matrix.cols());
+    matrix<double> residual_i = residual.block(start_i, 0, n_visits_i, 1);
+    matrix<double> gi_sqrt_root = G_sqrt.segment(start_i, n_visits_i).matrix().asDiagonal();
+    matrix<double> gi_simga_inv_chol = gi_sqrt_root * sigma_inv_chol;
+    matrix<double> xt_gi_simga_inv_chol = Xi.transpose() * gi_simga_inv_chol;
+    matrix<double> identity(n_visits_i, n_visits_i);
+    identity.setIdentity();
+    if (jackknife) {
+      identity = identity - xt_gi_simga_inv_chol.transpose() * beta_vcov_matrix * xt_gi_simga_inv_chol;
+    }
+    matrix<double> z =  xt_gi_simga_inv_chol * identity.inverse() * gi_simga_inv_chol.transpose() * residual_i;
+    meat = meat + z * z.transpose();
+  }
+  for (int r = 0; r < n_groups; r++) {
+    delete derivatives_by_group[r];
+  }
+  // beta_vcov already take gi into consideration;
+  matrix<double> ret = beta_vcov_matrix * meat * beta_vcov_matrix;
+  // Removed because this scale factor can be applied by user manually
+  // not important.
+  //if (jackknife) {
+  //  ret = ret * (n_subjects - 1) / n_subjects;
+  //}
```

Comment:
Why is this not important? Maybe we should still do that here if the method is described in the paper like that?

Reply:
the reason this scale parameter, though described in the paper, is not used in the `clubSandwich::vcovCR` for `gls`; to be consistent I removed that scale parameter but keep the commented code just in case we need it again; and in addition, the paper describes the empirical one with a additional $\frac{n}{n-1}$ scale parameter, which is also not used in `clubSandwich` and `SAS`. And in addition, if the weights of the observations are given, I wonder if this scaler still works. Having all these points I think it is good to remove the scale parameter for now